File: generated-code/csharp/codecraft/FileReadWrite/Model/BuildProperties.cs
namespace TransGenTest.Model
{
    /// <summary>
    /// Entity's build properties
    /// </summary>
    public struct BuildProperties
    {
        /// <summary>
        /// Valid new entity types
        /// </summary>
        public Model.EntityType[] Options { get; set; }
        /// <summary>
        /// Initial health of new entity. If absent, it will have full health
        /// </summary>
        public int? InitHealth { get; set; }
    
        public BuildProperties(Model.EntityType[] options, int? initHealth)
        {
            this.Options = options;
            this.InitHealth = initHealth;
        }
    
        /// <summary> Read BuildProperties from reader </summary>
        public static BuildProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new BuildProperties();
            result.Options = new Model.EntityType[reader.ReadInt32()];
            for (int optionsIndex = 0; optionsIndex < result.Options.Length; optionsIndex++)
            {
                result.Options[optionsIndex] = EntityTypeHelper.ReadFrom(reader);
            }
            if (reader.ReadBoolean())
            {
                result.InitHealth = reader.ReadInt32();
            } else
            {
                result.InitHealth = null;
            }
            return result;
        }
    
        /// <summary> Write BuildProperties to writer </summary>
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Options.Length);
            foreach (var optionsElement in Options)
            {
                writer.Write((int) (optionsElement));
            }
            if (!InitHealth.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                writer.Write(InitHealth.Value);
            }
        }
    
        /// <summary> Get string representation of BuildProperties </summary>
        public override string ToString() {
            string stringResult = "BuildProperties { ";
            stringResult += "Options: ";
            stringResult += "[ ";
            int optionsIndex = 0;
            foreach (var optionsElement in Options)
            {
                if (optionsIndex != 0) {
                    stringResult += ", ";
                }
                stringResult += optionsElement.ToString();
                optionsIndex++;
            }
            stringResult += " ]";
            stringResult += ", ";
            stringResult += "InitHealth: ";
            if (!InitHealth.HasValue)
            {
                stringResult += "null";
            } else
            {
                stringResult += InitHealth.Value.ToString();
            }
            stringResult += " }";
            return stringResult;
        }
    }
}

File: generated-code/pascal/codecraft/FileReadWrite/Model/BuildProperties.pas
unit BuildProperties;

{$mode delphi}{$H+}

interface

uses
    EntityType in 'Model/EntityType.pas',
    Nullable,
    Stream,
    SysUtils;

type
    // Entity's build properties
    TBuildProperties = class
        // Valid new entity types
        options: TArray<TEntityType>;
        // Initial health of new entity. If absent, it will have full health
        initHealth: TNullable<Int32>;
        constructor Create(options: TArray<TEntityType>; initHealth: TNullable<Int32>);
        // Read BuildProperties from input stream
        class function ReadFrom(stream: TStream): TBuildProperties; static;
        // Write BuildProperties to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TBuildProperties.Create(options: TArray<TEntityType>; initHealth: TNullable<Int32>);
begin
    self.options := options;
    self.initHealth := initHealth;
end;

class function TBuildProperties.ReadFrom(stream: TStream): TBuildProperties;
var initHealth: TNullable<Int32>;
var initHealthValue: Int32;
var options: TArray<TEntityType>;
var optionsElement: TEntityType;
var optionsIndex: Int32;
begin
    options := TArray<TEntityType>.Create;
    SetLength(options, stream.ReadInt32);
    for optionsIndex := 0 to Length(options) - 1 do begin
        optionsElement := TEntityType(stream.ReadInt32);
        options[optionsIndex] := optionsElement;
    end;
    if stream.ReadBoolean then begin
        initHealthValue := stream.ReadInt32;
        initHealth := initHealthValue;
    end else
        initHealth := nil;
    result := TBuildProperties.Create(options, initHealth);
end;

procedure TBuildProperties.WriteTo(stream: TStream);
var initHealthValue: Int32;
var optionsElement: TEntityType;
begin
    stream.WriteInt32(Length(options));
    for optionsElement in options do begin
        stream.WriteInt32(ord(optionsElement));
    end;
    if initHealth.HasValue then begin
        stream.WriteBoolean(true);
        initHealthValue := initHealth.Value;
        stream.WriteInt32(initHealthValue);
    end else
        stream.WriteBoolean(false);
end;

function TBuildProperties.ToString: ansistring;
var initHealthValue: Int32;
var optionsElement: TEntityType;
var optionsElementName: String;
var optionsIndex: Int32;
begin
    result := 'BuildProperties {';
    result += 'options=';
    result += '[';
    for optionsIndex := 0 to Length(options) - 1 do begin
        if optionsIndex <> 0 then
            result += ', ';
        optionsElement := options[optionsIndex];
        WriteStr(optionsElementName, optionsElement);
        result += optionsElementName;;
    end;
    result += ']';
    result += ', ';  
    result += 'initHealth=';
    if initHealth.HasValue then begin
        initHealthValue := initHealth.Value;
        result += IntToStr(initHealthValue);
    end else
        result += 'nil';
    result += '}';
end;

end.